src/ServiceControl.Monitoring/Timings/TimingsStore.cs

```diff
@@ -0,0 +1,141 @@
+﻿namespace ServiceControl.Monitoring.Timings
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using Metrics.Raw;
+
+    public abstract class TimingsStore
+    {
+        ConcurrentDictionary<string, ConcurrentDictionary<DateTime, MeasurementInterval>> timings = 
+            new ConcurrentDictionary<string, ConcurrentDictionary<DateTime, MeasurementInterval>>();
+
+        public void Store(string endpointName, LongValueOccurrences message, DateTime now)
+        {
+            var endpointData = timings.GetOrAdd(endpointName, _ => new ConcurrentDictionary<DateTime, MeasurementInterval>());
+
+            for (var i = 0; i < message.Ticks.Length; i++)
+            {
+                var date = new DateTime(message.BaseTicks + message.Ticks[i], DateTimeKind.Utc);
```

Comment:
If instead of deserializing data to a `LongValueOccurrences` and process it in here, we introduce a simple array of measurements being just an array of {Time, Value} measurements, we could do this work in the deserialization. Moreover, we could use pooling for these arrays and simply add this to a list of measurements for specific `endpointData`. This would lead to tranforming `endpointData` from a concurrent dictionary to a simple linked list of chunks of measurements (one chunk per incoming message).